tpie/pipelining/node.cpp

```diff
@@ -170,30 +163,31 @@ void node::add_dependency(const node & dest) {
 	add_dependency(dest.token);
 }
 
-void node::set_minimum_memory(memory_size_type minimumMemory) {
-	switch (get_state()) {
-	case STATE_IN_PROPAGATE:
-	case STATE_AFTER_PROPAGATE:
-	case STATE_FRESH:
-	case STATE_IN_PREPARE:
-		break;
-	default:
-		throw call_order_exception("set_minimum_memory");
-	}
-	m_parameters.minimumMemory = minimumMemory;
-}
 
-void node::set_maximum_memory(memory_size_type maximumMemory) {
-	switch (get_state()) {
-	case STATE_IN_PROPAGATE:
-	case STATE_AFTER_PROPAGATE:
-	case STATE_FRESH:
-	case STATE_IN_PREPARE:
-		break;
-	default:
-		throw call_order_exception("set_maximum_memory");
+#define PARAM_SETTER(setter_name, param_type, param_name) \
+	void node::setter_name(param_type param_name) { \
+		switch (get_state()) { \
+		case STATE_IN_PROPAGATE: \
+		case STATE_AFTER_PROPAGATE: \
+		case STATE_FRESH: \
+		case STATE_IN_PREPARE: \
+			break; \
+		default: \
+			throw call_order_exception(#setter_name); \
+		} \
+		m_parameters.param_name = param_name; \
 	}
-	m_parameters.maximumMemory = maximumMemory;
+
+PARAM_SETTER(set_minimum_file_descriptors, int, minimumFileDescriptors);
+PARAM_SETTER(set_maximum_file_descriptors, int, maximumFileDescriptors);
+PARAM_SETTER(set_file_descriptor_fraction, double, fileDescriptorFraction);
+
+PARAM_SETTER(set_minimum_memory, memory_size_type, minimumMemory)
+PARAM_SETTER(set_maximum_memory, memory_size_type, maximumMemory)
+PARAM_SETTER(set_memory_fraction, double, memoryFraction);
+
+void node::set_available_file_descriptors(int availableFileDescriptors) {
+	m_availableFileDescriptors = availableFileDescriptors;
 }
```

Comment:
An advantage of using a macro over a template is that we can use stringification to automatically get the function name: https://github.com/thomasmoelhave/tpie/pull/207/commits/41d6017f355a6c834c5f9caee4b2522ec0af7a95#diff-2cacb0a360fd6d9d9638fba375fcfd43R169